return fake create query when executing `show create table` on system's tables.

cmsxbc opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
return  fake create query when executing `show create table` on system's tables.

Detailed description / Documentation draft:
Current exception thrown when executing `SHOW CREATE TABLE system.users` is `Cannot open file`. It confused a lot of users that first time tried to get structure of system tables. Such as #29283, and there is no document mention it.
So this pr contains:
1. Throw exactly exception that some tables cannot be the target of `SHOW CREATE TABLE`.
2. Mention it in the document.

After discuss,  decided to show a fake query for these tables.

cmsxbc committed:
clarify that cannot show create table of system's tables

cmsxbc committed:
clarify that cannot show create table of system's tables in the document

alexey-milovidov commented:
Ok.

What is confusing me is that getting CREATE TABLE statement is the responsibility of IDatabase, not IStorage.
So, the method `supportsShowCreateTable` in IStorage is confusing. Let's ask @tavplubix about ideas...

cmsxbc commented:
@alexey-milovidov Because `system` is just a normal `DatabaseOnDisk`  and user can create any table under `system`. So we shall not disable it on IDatabase.
By the way, I will bypass style check because `SHOW CREATE TABLE` cannot use `database = currentDatabase()/$CLICKHOUSE_DATABASE` condition, as the check expected [check style result](https://clickhouse-test-reports.s3.yandex.net/31391/49abeb14b4dab8dade5a9bdae5a3e1a980f58f4e/style_check/style_output.txt)

cmsxbc committed:
bypass test style-check for required `database=currentDatabase()` because `show create table` do not have condition

tavplubix commented:
Let's rename `supportsShowCreateTable()` to `isSystemStorage()` or something like that.

cmsxbc committed:
rename supportsShowCreateTable to isSystemStorage

vitlibar commented:
I think maybe `SHOW CREATE TABLE` can show some fake `CREATE TABLE` statement for system tables, for example:

```
:) SHOW CREATE TABLE system.one

CREATE TABLE system.one (dummy UInt8) ENGINE=System
```

What do you think?

tavplubix commented:
> I think maybe `SHOW CREATE TABLE` can show some fake `CREATE TABLE` statement for system tables, for example:
> 
> ```
> :) SHOW CREATE TABLE system.one
> 
> CREATE TABLE system.one (dummy UInt8) ENGINE=System
> ```
> 
> What do you think?

It's good idea, we can create `ASTColumns` from `ColumnsDescription`, take engine name from `IStorage::getName()` and manually construct create query for system table. But we still need something like `IStorage::isSystemStorage()` method for this. Or we can do it for all tables and always show "fake" create query if metadata file does not exist, but it's questionable, because such behavior may hide some errors or bugs (but maybe we can add some assertions).

cmsxbc commented:
I think it's not a good idea to do it for all tables.The result of `SHOW CREATE TABLE` should be reusable, I always get the create query from one server and apply it to another server. It's confused if the statement from `SHOW CREATE TABLE` cannot be executed, unless I know exactly the statement is fake.

vitlibar commented:
> I think it's not a good idea to do it for all tables.The result of `SHOW CREATE TABLE` should be reusable, I always get the create query from one server and apply it to another server. It's confused if the statement from `SHOW CREATE TABLE` cannot be executed, unless I know exactly the statement is fake.

You will able to understand that the table is internal from the words `ENGINE=System` in its definition and knowing that the table engine `System` cannot be used to create tables. Or we could make the fact that the table is system more prominent by starting our "fake create statement" with `CREATE SYSTEM TABLE ...` if `ENGINE=System` doesn't seem to be visible enough. My point is that some people might use `SHOW CREATE TABLE` to get the information about the columns, their names and types, and not to recreate the table.

cmsxbc committed:
build fake create query for show create table of system's tables

cmsxbc commented:
@vitlibar @tavplubix I have implemented one

cmsxbc committed:
fix tests for show create table on system's table:

1. only build fake create query when cannot get real query
2. only test rocksdb when it's enabled.

cmsxbc committed:
fix test for show create table system.tables

cmsxbc committed:
Merge branch 'ClickHouse:master' into clarify-cannot-show-create-table-of-system

cmsxbc committed:
fix test for ASTCreateQuery set table and database

tavplubix commented on src/Databases/DatabaseOnDisk.cpp:
```diff
@@ -673,6 +679,69 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromMetadata(const String & database_metada
+                ParserDataType parser;
+                if (!parser.parse(pos, ast_type, expected))
+                {
+                    /// this should never occur, shall we change to use assert?
+                    if (throw_on_error)
+                        throw Exception(ErrorCodes::CANNOT_GET_CREATE_TABLE_QUERY, "Cannot parser metadata of {}.{}", backQuote(getDatabaseName()), backQuote(table_name));
```
```suggestion
                    if (throw_on_error)
                        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot parser metadata of {}.{}", backQuote(getDatabaseName()), backQuote(table_name));
```

tavplubix replied:
Exceptions with `LOGICAL_ERROR` code work like assert in debug builds and like ordinary exception in release builds

tavplubix commented on src/Databases/DatabaseOnDisk.cpp:
```diff
@@ -673,6 +679,69 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromMetadata(const String & database_metada
+        ast_storage->set(ast_storage->engine, ast_engine);
+        create_table_query->set(create_table_query->storage, ast_storage);
+    }
+    /// setup create table query columns info.
```
Btw, we already have similar code [here](https://github.com/ClickHouse/ClickHouse/blob/7b5954b736465aa059803139a3870aeab3d2d327/src/Databases/MySQL/DatabaseMySQL.cpp#L116) and [here](https://github.com/ClickHouse/ClickHouse/blob/d76976e272268f76bb469d611aca66acd9dd1cbf/src/Databases/SQLite/DatabaseSQLite.cpp#L173). Is it possible to extract common code to some function or static method?

cmsxbc replied:
Ok, I will have a try.

cmsxbc committed:
change to LOGICAL_ERROR when cannot get in_memory metadata

Co-authored-by: tavplubix <tavplubix@gmail.com>

cmsxbc committed:
implement common getCreateQueryFromStorage

tavplubix commented:
Integration tests - known issue with iptables rules
Performance - #31530
Stress test (debug, actions) - #31531
